### ISSUE TITLE
Update `kubectl` documentation for Server Side Apply

### ISSUE THREAD

**Issue by sftim** [kubernetes/website#39145]
**This is a Feature Request**

**What would you like to be added**
In https://github.com/kubernetes/website/issues/39108#issuecomment-1407679439 I explained that I think we missed documenting a key change to `kubectl` that was made when we implemented [server-side apply](https://github.com/kubernetes/enhancements/issues/555).

> &hellip;the input to `kubectl apply` is no longer defined to be a _manifest_: it's actually an _apply configuration_, and all valid manifests are also valid apply configurations. However, `kubectl apply` defaults to validating that the supplied _apply configuration_ is a valid manifest. You can then turn off that validation by using `--validate=false`.

&hellip;

> I think we should clearly document which kubectl subcommands work with manifests and which with apply configurations.

&hellip;

> A few subcommands and plugins to consider:
> 
>  - `kubectl create`
>  - `kubectl replace`
>  - `kubectl diff`
>  - `kubectl convert`

**Why is this needed**

(If I've understood correctly) `kubectl apply` does accept YAML manifests as its input, but this is actually a side effect of its actual input format being a new thing called an _apply configuration_, which is a subset of the fields from a full manifest.

**Comments**
- Let's use this issue, not #39108, to track any documentation changes that are needed.
- We might also want to update the kubectl docs and source code to account for this change, again provided that I've understood it right.
  
/wg api-expression
/language en

/lifecycle frozen
We should triage this issue, no matter how long that process takes.

**Comment by apelisse** [comment]
> I think --validate is from an era when the server did less checking and gave less helpful error messages. I'm not sure it's useful at all anymore.

And it's being replaced by [UnknownFieldValidation](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2885-server-side-unknown-field-validation) which should go GA in the next release, and completely deprecate client-side `--validate` in the next release.

**Comment by sftim** [comment]
I think at some early point in the evolution of Kubernetes we decided that you could specify a partial thing rather than a full, valid manifest, but we never properly documented what these partial things are.

It doesn't really matter exactly when we missed the opportunity; we still need to do it.

**Comment by lavalamp** [comment]
I'm not saying we shouldn't fix the documentation, just trying to make it clear that it isn't a CSA/SSA difference and the documentation should be the same for both.

**Comment by sftim** [comment]
Indeed. The input to `kubectl apply` is a [something] regardless of whether `kubectl` then does an apply client-side or server-side. And it's the same [something].

**Comment by lavalamp** [comment]
I think I called it a "Partially Specified Object" in design docs; "ApplyConfiguration" is language from the fancy go client, I probably wouldn't use that language in user facing docs?

**Comment by sftim** [comment]
If this is a valid gap in docs, let's triage it accepted and work on a sketch for getting some changes in.

**Comment by antaloala** [comment]
Hmm ... is there anything with respect to SSA (and these "Partially Specified Objects/manifests") that has been changed from 1.25 to 1.26 ?
The following test worked in 1.25:

``kubectl create namespace test``

``# yaml manifest below is a full Deployment one``
``kubectl apply -f deployment-busybox.yaml -n test --server-side``

``# yaml manifest below is a partial one, only containing the .spec.replicas setting``
``# in order to modify it (keeping .apiversion, .kind and .metadata.name equal to the previous yaml)`` 
``kubectl apply -f ssa-patch-replicas.yaml -n test --server-side --validate=false`` 

I now get schema complaints from API server after having upgraded to 1.26, getting it independently of setting ``--validation=true`` or ``--validation=ignore``
See the trace report when using ``-v=8``

#### When ``--validate=ignore``
[...]
I0201 13:03:08.642957  157928 request.go:1154] Request Body: {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"name":"busybox","namespace":"test"},"spec":{"replicas":2}}
I0201 13:03:08.643073  157928 round_trippers.go:463] `PATCH https://172.16.0.3:6443/apis/apps/v1/namespaces/test/deployments/busybox?fieldManager=kubectl&`**fieldValidation=Ignore&force=false**
[...]
The Deployment "busybox" is invalid: 
* spec.selector: Required value
* spec.template.metadata.labels: Invalid value: map[string]string(nil): `selector` does not match template `labels`
* spec.template.spec.containers: Required value
* spec.selector: Invalid value: "null": field is immutable


#### When ``--validate=true``
[...]
I0201 12:56:16.039715  153918 request.go:1154] Request Body: {"apiVersion":"apps/v1","kind":"Deployment","metadata":{"name":"busybox","namespace":"test"},"spec":{"replicas":2}}
I0201 12:56:16.039848  153918 round_trippers.go:463] `PATCH https://172.16.0.3:6443/apis/apps/v1/namespaces/test/deployments/busybox?fieldManager=kubectl&`**fieldValidation=Strict&force=false**
[...]
The Deployment "busybox" is invalid: 
* spec.selector: Required value
* spec.template.metadata.labels: Invalid value: map[string]string(nil): `selector` does not match template `labels`
* spec.template.spec.containers: Required value
* spec.selector: Invalid value: "null": field is immutable

#### kubectl version --short
Client Version: v1.26.0
Kustomize Version: v4.5.7
Server Version: v1.26.0

Should I raise an issue?

**Comment by sftim** [comment]
https://github.com/kubernetes/website/issues/39145#issuecomment-1411974714 feels like a separate thing and not directly related to https://github.com/kubernetes/website/issues/39145. However, having unclear or incomplete docs makes it hard to know for sure.

**Comment by sftim** [comment]
I edited https://github.com/kubernetes/website/issues/39145#issuecomment-1411974714 for formatting

**Comment by antaloala** [comment]
> [#39145 (comment)](https://github.com/kubernetes/website/issues/39145#issuecomment-1411974714) feels like a separate thing and not directly related to #39145. However, having unclear or incomplete docs makes it hard to know for sure.

I am mentioning here because now I do not know if 1.25 behavior was ok and 1.26 is then wrong ... or the opposite(?). When I raised #39108 I assumed 1.25 was the desired behavior (I raised it without testing in 1.26). But if being the opposite: the "partial manifest" are not possible .. what would not make any sense in the way I understand SSA.

@lavalamp, @apelisse  could you comment? Should I raise an issue/bug towards 1.26 on SSA schema-validation behavior on partial manifests?

**Comment by apelisse** [comment]
OK, let me try to clarify the situation.
- Objects validation is not optional in Kubernetes, the apiserver will ALWAYS refuse invalid objects.
- Objects that have extraneous (unknown) fields, or redundant fields, will be valid since the extra-fields are automatically (and silently) dropped by the apiserver
- Kubectl used to validate these extra-fields using the OpenAPI (and the --validate flag)
- We've moved the validation of extra-fields from kubectl to the apiserver (also controlled with the --validate field, pay attention to the menu because the options have changed). One can ask the apiserver to drop with a warning (new default), drop silently (former default), or reject the request

**Comment by apelisse** [comment]
cc @kevindelgado

**Comment by lavalamp** [comment]
@antaloala did you also retain namespace? This probably isn't the right place for this discussion, please open an issue (and show exact manifests if possible) in the main repo.

**Comment by antaloala** [comment]
Thanks @apelisse. The validation part is clear now (I misunderstood it). Not really relevant for this SSA discussion on partial manifests being/not being valid input yamls when using SSA (my mistake when mixing both).

After playing a little more with the partial manifests I think I finally got it (I also made some mistakes in my previous reasoning):

- I repeated the kubectl sequence listed above (in https://github.com/kubernetes/website/issues/39145#issuecomment-1411974714) but this time the initial "deployment.yaml" does not contain any .spec.replicas setting (so the default "1" value applies but the .spec.replicas filed is not own by the used field-manager)
- Now I run the second kubectl --server-side where using a partial manifest ONLY setting the .spec.replicas AND using a different filed-manager => it works

But if trying to do the second step above with the same field-manager I used in the original "creation" of the deployment object => it fails (saying I am missing fields)

So I understand SSA is so well implemented that if a field-manager becomes owner for some fields => following SSA PATH request must keep, in the partial manifests, all thise currently owned fields + new optional ones.

Is it like that?

**Comment by lavalamp** [comment]
Yes, if you use the same field manager, apiserver believes you are trying to delete all the other fields owned by that manager when you omit them.

**Comment by antaloala** [comment]
> Yes, if you use the same field manager, apiserver believes you are trying to delete all the other fields owned by that manager when you omit them.

Aha, that makes all sense (as I was trying to remove fields that are "required" in the deployments schema .. the reason for the "schema error" I was getting from the API Server.

So if a manager becomes owner for a "required" yaml/json field (according to the schema governing the GVK for the involved API object) => that manager needs to keep that field in the following "partial manifests" it sent in following SSA PATCH requests (it not other manager takes the ownership), right?

let me say the SSA implementation is really awesome (it is even better than I initially understood).

**Comment by sftim** [comment]
> @sftim this info would be really useful (I think) for the SSA missing info in k8s doc. Do you agree?

Please file a new issue. Then, I'll mark the comments in _this_ issue as off-topic and hide them, so that this issue stays on track.

**Comment by lavalamp** [comment]
"fully specified intent" seems OK to me.

**Comment by antaloala** [comment]
@sftim note in my previous comment I am not saying anything against improving current k8s doc and/or kubectl documentation with respect to SSA missing info. When saying "we do not really need any other new issue" I am only referring to that new issue you suggested to file around how SSA implementation behaves when a manager, in a following SSA Apply request, omit fields for which it is the single current owner (saying we do not need to file an issue on this SSA details as current k8s docs on SSA already describe all this).
